### PR TITLE
Add configuration options to change the upstream URL

### LINKED ISSUE #61: Add configuration options to change the upstream URL
In https://github.com/luuuis/hass_wibeee/issues/59 you suggested me to use the [Local Push support](https://github.com/luuuis/hass_wibeee#configuring-local-push-optional-advanced), but I couldn't use it since my provider uses another upstream other than nest and they have their own app and so on. But I noticed that pulling all the time makes the wibeee device unresponsible from time to time and it doesn't happen when I stop pulling, so I want to try the 'Local Push support', as it should give no extra load to the device.

I think this is a great feature for all the people that does not use the official 'http://nest-ingest.wibeee.com' upstream.

### PR DESCRIPTION
**Pull request by edurenye**
Allows selecting a different Cloud service to upload. Also introduces Local-only mode for those who want to use Local Push without sending the data off to any Cloud service.

Wibeee options

When running the new version for the first time the `ConfigEntry` is migrated to version 2 to support the new configuration select box.

### CODE REVIEW

**Comment by edurenye** [comment]
I started to work on this issue but I'm getting an error that I can't figure out.

`2023-04-15 21:04:31.763 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up wibeee platform for sensor
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 304, in _async_setup_platform
await asyncio.shield(task)
File "/config/custom_components/wibeee/sensor.py", line 203, in async_setup_entry
await get_nest_proxy(
TypeError: get_nest_proxy() takes 1 positional argument but 2 were given`

But the code actually takes 2 positional arguments:
`async def get_nest_proxy(`
`        hass: HomeAssistant,`
`        entry: ConfigEntry`
`) -> NestProxy:`

File: custom_components/wibeee/nest.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 from homeassistant.components.network import async_get_source_ip
 from homeassistant.components.network.const import PUBLIC_TARGET_IP
+from homeassistant.config_entries import ConfigEntry
```
Comment (luuuis) [inline custom_components/wibeee/nest.py:7]: Let's please keep this module independent of the config flow. That is, pass in the primitive str/int values instead of the ConfigEntry.

File: custom_components/wibeee/nest.py [hunk 2]
```diff
@@ -38,11 +45,12 @@ def dispatch(self, mac_addr: str, pushed_data: Dict):
 @singleton.singleton("wibeee_nest_proxy")
 async def get_nest_proxy(
         hass: HomeAssistant,
-        upstream='http://nest-ingest.wibeee.com',
-        local_port=8600,
+        entry: ConfigEntry
```
Comment (luuuis) [inline custom_components/wibeee/nest.py:48]: This is a `@singleton`, which means that there is only a single instance for all of Home Assistant. The decorator expects that the factory def must only take a single param of type `HomeAssistant` (you can [see how it's implemented here](https://github.com/luuuis/home-assistant-core/blob/dev/homeassistant/helpers/singleton.py#L38-L52)) and that is the reason why you're getting `TypeError: get_nest_proxy() takes 1 positional argument but 2 were given`..

Basically it will not be possible to have a singleton per `ConfigEntry` and probably you don't want an instance per `ConfigEntry` either (too many ports used and unnecessary).

File: custom_components/wibeee/sensor.py [hunk 3]
```diff
@@ -162,9 +167,12 @@ async def fetching_data(now=None):
     return async_track_time_interval(hass, fetching_data, scan_interval)
 
 
-async def async_setup_local_push(hass: HomeAssistant, device, sensors: list['WibeeeSensor']):
+async def async_setup_local_push(hass: HomeAssistant, entry: ConfigEntry, device, sensors: list['WibeeeSensor']):
     mac_address = device['macAddr']
-    nest_proxy = await get_nest_proxy(hass)
+    nest_proxy = await get_nest_proxy(
```
Comment (luuuis) [inline custom_components/wibeee/sensor.py:172]: Ok so as I mentioned in the other comment this approach is not going to work. But you can see below we have a call and that would be the place to add additional information specific to this device (MAC address).

You could update it to work like this:

```python
nest_proxy.register_device(mac_address, upstream=entry.get(...), listener=on_pushed_data)
```

The `nest` module will then need some refactoring to forward to the configured upstream for that MAC address. Does that make sense? (but don't start just yet as I would like to explore using another approach, which is to have a dropdown selection menu for the upstream)

**Review by luuuis — COMMENTED** [review]
Thanks for the PR. I agree with supporting different upstreams, the only thing I would like to explore is having a list of builtin upstreams that users can select, instead of providing text boxes.

Please have a look at my comments and let me know what you think.

File: custom_components/wibeee/nest.py (hunk 1 above)
Comment (edurenye) [reply custom_components/wibeee/nest.py:7]: That was my first approach, but since I had that error about the parameters I changed it to that, just to try.
Also I wanted to add an advanced configuration option to add the new settings, but couldn't find how to do it :upside_down_face:

File: custom_components/wibeee/nest.py (hunk 2 above)
Comment (edurenye) [reply custom_components/wibeee/nest.py:48]: ok, I understand

File: custom_components/wibeee/sensor.py (hunk 3 above)
Comment (edurenye) [reply custom_components/wibeee/sensor.py:172]: It makes sense that the configuration is for device and not global, just in case there are wibeee devices that have different upstreams, would be strange I guess but possible.

Thank you very much for your help.

It's the first time I try to develop with home assistant, but I'm willing to learn and maybe in the future be able to create my own plugins or addons.

Comment (edurenye) [reply custom_components/wibeee/sensor.py:172]: Though I'm against the dropdown since there might be tons of upstreams, I saw in a comment somewhere that you mentioned a few upstreams, but non of those where the upstream that I need: my provider is a Spanish company called 'SolarProfit', and they use the endpoint 'http://wdata.solarprofit.es' and the port 8080, that is why I also made the upstream port configurable. What we could make is give a dropdown of options but with the chance of adding your own somehow.

I'll try to fix the change mentioned before and we can further discuss about how to approach the UI.

Comment (edurenye) [reply custom_components/wibeee/sensor.py:172]: Now I see that making the config by device is not what you meant :upside_down_face: 
Forget about what I said there.

Comment (luuuis) [reply custom_components/wibeee/sensor.py:175]: Configuration must be per device simply because that is how Home Assistant config flow works. I don't think there's any option (outside of YAML) to make "global" configurations that apply to multiple config entries. When you click "Configure" on an integration those options are for one integration only (one Wibeee device in this case).

https://developers.home-assistant.io/docs/config_entries_config_flow_handler
https://developers.home-assistant.io/docs/config_entries_options_flow_handler

I'm happy to be proven wrong though.

Comment (edurenye) [reply custom_components/wibeee/sensor.py:175]: Yes, sorry I was confused because the lines appear listed as devices, I'm still trying to figure things out of how home assistant works.

I was trying to use the 'show_advanced_options' but I'm not sure how does it work and I can't find it in the documentation.

Also I found out how to use the dropdown, I just need to know where to find a full list of servers, I just found this:
'http://nest-ingest.wibeee.com',
'http://wdata.solarprofit.es:8080'
'http://wibeee.smilics.com'
'http://monitorconsumo.iberdrola.es'
'http://wattius.mirubee.com'
'http://www.smilics.com:8080'

And those might not exist anymore or have changed, that is why I think we should leave the option to add it written, but maybe instead of having to fields for URL and port it is better to combine into just one field.

Comment (luuuis) [reply custom_components/wibeee/sensor.py:175]: > I was trying to use the 'show_advanced_options' but I'm not sure how does it work and I can't find it in the documentation.

It probably has something to do with the [per-user setting where you can specify that you are "advanced"](https://www.home-assistant.io/blog/2019/07/17/release-96/#advanced-mode). In which case I would probably not use it.

Could you not do the following in the options flow?
1. Keep the existing checkbox on the first step.
2. When that is submitted then look at the submitted value, if disabled then call `self.async_create_entry` as before but if it is enabled then show a new step with the extra configs that you are adding.

Comment (edurenye) [reply custom_components/wibeee/sensor.py:175]: I wish it would appear when checking the checkbox, but not sure if that is possible, couldn't find a way to do it in the [voluptuous documentation](https://alecthomas.github.io/voluptuous/docs/_build/html/voluptuous.html)

IMHO having to save the form before getting more options might confuse the users.

Regarding the dropdown, do you still think I should add a dropdown? If so, which upstreams should I add?

Comment (edurenye) [reply custom_components/wibeee/sensor.py:174]: I made it as radio buttons, I was following and example that was showing up as a drop-down, not sure what I did wrong, but it looks good also with the radio buttons, I left it with the URL but I could also add some text more user friendly like:
Screenshot from 2023-04-26 12-50-28

Comment (luuuis) [reply custom_components/wibeee/sensor.py:174]: Thanks. Is the PR ready for me to take a look and test?

Do you need the local port to be configurable as in this screenshot?

Comment (edurenye) [reply custom_components/wibeee/sensor.py:174]: Hi, sorry for the laste answer. Is not that I need it configurable, I just thought would be nice in case some other application/plugin/component already is using that same port.

**Comment by luuuis** [comment]
@edurenye I have pushed a few commits. Please see the [updated description](https://github.com/luuuis/hass_wibeee/pull/62#issue-1669545716) for how it is configured now.

Are you able to test this locally? Preliminary tests are working for me but since the clamps on my Wibeee are measuring devices that are not often the device ends up sending data only every 1 minute anyway.

**Comment by edurenye** [comment]
> @edurenye I have pushed a few commits. Please see the [updated description](https://github.com/luuuis/hass_wibeee/pull/62#issue-1669545716) for how it is configured now.
> 
> Are you able to test this locally? Preliminary tests are working for me but since the clamps on my Wibeee are measuring devices that are not often the device ends up sending data only every 1 minute anyway.

Nice, I'll try it out, thank you!

**Comment by edurenye** [comment]
Well, I'll wait to test the pushing since some times it malfunctions when changing the configuration and I need to reset the wibeee, wanna be there phisically to be able to reset, I contacted SolarProfit regarding this issue and they said a technician would contact me but it's been a while and the haven't contacted me, so I'll have to call them again.
